sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala

```diff
@@ -128,11 +130,39 @@ class QueryExecution(
     case _ => "command"
   }
 
+  private def assertNoStreamSourceMarkerNode(p: LogicalPlan): Unit = {
+    // In UnsupportedOperationChecker.checkForBatch, we check if the plan has any streaming node.
+    // That is more aggressive than just checking the marker node for streaming source which is
+    // yet to be materialized. We'd like to be a bit conservative here since this is the exact
+    // problematic case we figured out.
+    p.foreach {
+      case _: StreamingRelation | _: StreamingRelationV2 |
+           _: StreamingExecutionRelation | _: StreamingDataSourceV2ScanRelation =>
+        val msg = "Queries with streaming sources must be executed with writeStream.start()"
+        // This is exactly the same with UnsupportedOperationChecker.checkForBatch.
+        // TODO: Classify and issue a new error class, along with UnsupportedOperationChecker.
+        throw new ExtendedAnalysisException(
+          new AnalysisException(
+            errorClass = "_LEGACY_ERROR_TEMP_3102",
+            messageParameters = Map("msg" -> msg)),
+          plan = p)
+
+      case _ =>
+    }
+  }
+
   private def eagerlyExecuteCommands(p: LogicalPlan) = {
     def eagerlyExecute(
         p: LogicalPlan,
         name: String,
         mode: CommandExecutionMode.Value): LogicalPlan = {
+      // Since we are about to execute the plan, the plan shouldn't have a marker node to be
+      // materialized during microbatch planning. If the plan has a marker node, it is highly
+      // likely that users put streaming sources in a batch query.
+      // This case brings problem before reaching the check in UnsupportedOperationChecker,
```

Comment:
I think the eager command execution is similar to a normal query execution, why can't it hit `UnsupportedOperationChecker`?

Reply:
Because `assertSupported` is not called for eagerly executed command?

Reply:
This is actually coupled with "explain". The test in QueryExecutionSuite ends up with StackOverflowError. I'll leave a PR comment what happens in the test suite.

Reply:
https://github.com/apache/spark/pull/50015/files#r1963056691